### PR TITLE
feat(query): support h3_to_geo/h3_to_geo_boundary/h3_k_ring function

### PR DESCRIPTION
**Pull request by nange**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. Support `h3_to_geo`, `h3_to_geo_boundary`, `h3_k_ring` function. 
2. Fix the panic issue in the `geo_to_h3` function when the resolution parameter exceeds 15.

- Related #6390

### CODE REVIEW

**Comment by ariesdevil** [comment]
Hi @nange , as h3 is an independent geo system, it's better place all h3 related functions to a new file like `geo_h3.rs`

**Comment by nange** [comment]
Hi @ariesdevil . I encountered an issue while running `cargo test` in my local env.  The error message is as follows:
```rust
error: expected one of `:`, `;`, `=`, `@`, or `|`, found `-`
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:72749:15
      |
72749 |     let mut f_-1: Option<NoSuchObjectException> = None;
      |               ^ expected one of `:`, `;`, `=`, `@`, or `|`

error[E0463]: can't find crate for `ordered_float`
 --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:5:1
  |
5 | extern crate ordered_float;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error[E0463]: can't find crate for `try_from`
 --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:7:1
  |
7 | extern crate try_from;
  | ^^^^^^^^^^^^^^^^^^^^^^ can't find crate

   Compiling common-management v0.1.0 (/home/nange/rust/github.com/nange/databend/src/query/management)
error[E0428]: the name `GetTablesResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:48813:1
      |
16849 | pub struct GetTablesResult {
      | -------------------------- previous definition of the type `GetTablesResult` here
...
48813 | struct GetTablesResult {
      | ^^^^^^^^^^^^^^^^^^^^^^ `GetTablesResult` redefined here
      |
      = note: `GetTablesResult` must be defined only once in the type namespace of this module

error[E0428]: the name `GetTableResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:49583:1
      |
16685 | pub struct GetTableResult {
      | ------------------------- previous definition of the type `GetTableResult` here
...
49583 | struct GetTableResult {
      | ^^^^^^^^^^^^^^^^^^^^^ `GetTableResult` redefined here
      |
      = note: `GetTableResult` must be defined only once in the type namespace of this module

error[E0428]: the name `AddPartitionsResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:51547:1
      |
10061 | pub struct AddPartitionsResult {
      | ------------------------------ previous definition of the type `AddPartitionsResult` here
...
51547 | struct AddPartitionsResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^ `AddPartitionsResult` redefined here
      |
      = note: `AddPartitionsResult` must be defined only once in the type namespace of this module

error[E0428]: the name `GetFileMetadataByExprResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:69280:1
      |
15722 | pub struct GetFileMetadataByExprResult {
      | -------------------------------------- previous definition of the type `GetFileMetadataByExprResult` here
...
69280 | struct GetFileMetadataByExprResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `GetFileMetadataByExprResult` redefined here
      |
      = note: `GetFileMetadataByExprResult` must be defined only once in the type namespace of this module

error[E0428]: the name `GetFileMetadataResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:69394:1
      |
15906 | pub struct GetFileMetadataResult {
      | -------------------------------- previous definition of the type `GetFileMetadataResult` here
...
69394 | struct GetFileMetadataResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `GetFileMetadataResult` redefined here
      |
      = note: `GetFileMetadataResult` must be defined only once in the type namespace of this module

error[E0428]: the name `PutFileMetadataResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:69508:1
      |
16046 | pub struct PutFileMetadataResult {
      | -------------------------------- previous definition of the type `PutFileMetadataResult` here
...
69508 | struct PutFileMetadataResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `PutFileMetadataResult` redefined here
      |
      = note: `PutFileMetadataResult` must be defined only once in the type namespace of this module

error[E0428]: the name `ClearFileMetadataResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:69622:1
      |
16192 | pub struct ClearFileMetadataResult {
      | ---------------------------------- previous definition of the type `ClearFileMetadataResult` here
...
69622 | struct ClearFileMetadataResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `ClearFileMetadataResult` redefined here
      |
      = note: `ClearFileMetadataResult` must be defined only once in the type namespace of this module

error[E0428]: the name `CacheFileMetadataResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:69736:1
      |
16300 | pub struct CacheFileMetadataResult {
      | ---------------------------------- previous definition of the type `CacheFileMetadataResult` here
...
69736 | struct CacheFileMetadataResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `CacheFileMetadataResult` redefined here
      |
      = note: `CacheFileMetadataResult` must be defined only once in the type namespace of this module

warning: variant `LOAD_DONE` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:107:3
    |
107 |   LOAD_DONE = 1,
    |   ^^^^^^^^^ help: convert the identifier to upper camel case: `LoadDone`
    |
    = note: `#[warn(non_camel_case_types)]` on by default

warning: variant `NOT_ACQUIRED` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:214:3
    |
214 |   NOT_ACQUIRED = 4,
    |   ^^^^^^^^^^^^ help: convert the identifier to upper camel case: `NotAcquired`

warning: variant `SHARED_READ` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:249:3
    |
249 |   SHARED_READ = 1,
    |   ^^^^^^^^^^^ help: convert the identifier to upper camel case: `SharedRead`

warning: variant `SHARED_WRITE` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:250:3
    |
250 |   SHARED_WRITE = 2,
    |   ^^^^^^^^^^^^ help: convert the identifier to upper camel case: `SharedWrite`

warning: variant `NO_TXN` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:358:3
    |
358 |   NO_TXN = 6,
    |   ^^^^^^ help: convert the identifier to upper camel case: `NoTxn`

warning: variant `SCHEMA_REGISTRY` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:432:3
    |
432 |   SCHEMA_REGISTRY = 2,
    |   ^^^^^^^^^^^^^^^ help: convert the identifier to upper camel case: `SchemaRegistry`

warning: variant `START_REVIEW` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:572:3
    |
572 |   START_REVIEW = 2,
    |   ^^^^^^^^^^^^ help: convert the identifier to upper camel case: `StartReview`

warning: variant `CHANGES_REQUIRED` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:573:3
    |
573 |   CHANGES_REQUIRED = 3,
    |   ^^^^^^^^^^^^^^^^ help: convert the identifier to upper camel case: `ChangesRequired`

warning: variant `ORC_SARG` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:685:3
    |
685 |   ORC_SARG = 1,
    |   ^^^^^^^^ help: convert the identifier to upper camel case: `OrcSarg`

warning: variant `TEST_CAPABILITY` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:717:3
    |
717 |   TEST_CAPABILITY = 1,
    |   ^^^^^^^^^^^^^^^ help: convert the identifier to upper camel case: `TestCapability`

warning: variant `INSERT_ONLY_TABLES` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-6fde24d37691ab35/out/hms_patched.rs:718:3
    |
718 |   INSERT_ONLY_TABLES = 2,
    |   ^^^^^^^^^^^^^^^^^^ help: convert the identifier to upper camel case: `InsertOnlyTables`

   Compiling common-storages-iceberg v0.1.0 (/home/nange/rust/github.com/nange/databend/src/query/storages/iceberg)
   Compiling common-sql v0.1.0 (/home/nange/rust/github.com/nange/databend/src/query/sql)
Some errors have detailed explanations: E0428, E0463.
For more information about an error, try `rustc --explain E0428`.
warning: `common-hive-meta-store` (lib) generated 11 warnings
error: could not compile `common-hive-meta-store` (lib) due to 11 previous errors; 11 warnings emitted
```
Do you know why?  How can I fix it?  

Btw: `make test` will be get the same result.

**Comment by ariesdevil** [comment]
Hi @nange , please rebase main then try `make clean` and `REGENERATE_GOLDENFILES=1 make test` again.

**Comment by nange** [comment]
@ariesdevil  
After rebase main, then do `make clean` and `REGENERATE_GOLDENFILES=1 make test` which got the same issue as above. 

And I also retry `REGENERATE_GOLDENFILES=1 cargo test --color=always --test it 'scalars::geo::test_geo' --manifest-path /home/nange/rust/github.com/nange/databend/src/query/functions/Cargo.toml -- --show-output` which got another issue as follows: 
```rust
  Compiling common-hashtable v0.1.0 (/home/nange/rust/github.com/nange/databend/src/common/hashtable)
   Compiling common-expression v0.1.0 (/home/nange/rust/github.com/nange/databend/src/query/expression)
   Compiling common-meta-app v0.1.0 (/home/nange/rust/github.com/nange/databend/src/meta/app)
   Compiling common-functions v0.1.0 (/home/nange/rust/github.com/nange/databend/src/query/functions)
error[E0277]: the trait bound `Tz: Serialize` is not satisfied
    --> src/meta/app/src/background/background_job.rs:90:10
     |
90   | #[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Default, Eq, PartialEq)]
     |          ^^^^^^^^^^^^^^^^ the trait `Serialize` is not implemented for `Tz`
...
95   |     pub scheduled_job_timezone: Option<chrono_tz::Tz>,
     |     ------------------------------------------------- required by a bound introduced by this call
     |
     = help: the following other types implement trait `Serialize`:
               &'a T
               &'a mut T
               ()
               (T0, T1)
               (T0, T1, T2)
               (T0, T1, T2, T3)
               (T0, T1, T2, T3, T4)
               (T0, T1, T2, T3, T4, T5)
             and 789 others
     = note: required for `std::option::Option<Tz>` to implement `Serialize`
note: required by a bound in `app_error::_::_serde::ser::SerializeStruct::serialize_field`
    --> /home/nange/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/ser/mod.rs:1901:12
     |
1895 |     fn serialize_field<T: ?Sized>(
     |        --------------- required by a bound in this associated function
...
1901 |         T: Serialize;
     |            ^^^^^^^^^ required by this bound in `SerializeStruct::serialize_field`

error[E0277]: the trait bound `Tz: Deserialize<'_>` is not satisfied
    --> src/meta/app/src/background/background_job.rs:95:33
     |
95   |     pub scheduled_job_timezone: Option<chrono_tz::Tz>,
     |                                 ^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `Tz`
     |
     = help: the following other types implement trait `Deserialize<'de>`:
               <&'a Path as Deserialize<'de>>
               <&'a [u8] as Deserialize<'de>>
               <&'a str as Deserialize<'de>>
               <() as Deserialize<'de>>
               <(T0, T1) as Deserialize<'de>>
               <(T0, T1, T2) as Deserialize<'de>>
               <(T0, T1, T2, T3) as Deserialize<'de>>
               <(T0, T1, T2, T3, T4) as Deserialize<'de>>
             and 820 others
     = note: required for `std::option::Option<Tz>` to implement `Deserialize<'_>`
note: required by a bound in `next_element`
    --> /home/nange/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/de/mod.rs:1729:12
     |
1727 |     fn next_element<T>(&mut self) -> Result<Option<T>, Self::Error>
     |        ------------ required by a bound in this associated function
1728 |     where
1729 |         T: Deserialize<'de>,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `SeqAccess::next_element`

error[E0277]: the trait bound `Tz: Deserialize<'_>` is not satisfied
    --> src/meta/app/src/background/background_job.rs:95:33
     |
95   |     pub scheduled_job_timezone: Option<chrono_tz::Tz>,
     |                                 ^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `Tz`
     |
     = help: the following other types implement trait `Deserialize<'de>`:
               <&'a Path as Deserialize<'de>>
               <&'a [u8] as Deserialize<'de>>
               <&'a str as Deserialize<'de>>
               <() as Deserialize<'de>>
               <(T0, T1) as Deserialize<'de>>
               <(T0, T1, T2) as Deserialize<'de>>
               <(T0, T1, T2, T3) as Deserialize<'de>>
               <(T0, T1, T2, T3, T4) as Deserialize<'de>>
             and 820 others
     = note: required for `std::option::Option<Tz>` to implement `Deserialize<'_>`
note: required by a bound in `next_value`
    --> /home/nange/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/de/mod.rs:1868:12
     |
1866 |     fn next_value<V>(&mut self) -> Result<V, Self::Error>
     |        ---------- required by a bound in this associated function
1867 |     where
1868 |         V: Deserialize<'de>,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `MapAccess::next_value`

error[E0277]: the trait bound `Tz: Deserialize<'_>` is not satisfied
  --> src/meta/app/src/background/background_job.rs:95:5
   |
95 |     pub scheduled_job_timezone: Option<chrono_tz::Tz>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `Tz`
   |
   = help: the following other types implement trait `Deserialize<'de>`:
             <&'a Path as Deserialize<'de>>
             <&'a [u8] as Deserialize<'de>>
             <&'a str as Deserialize<'de>>
             <() as Deserialize<'de>>
             <(T0, T1) as Deserialize<'de>>
             <(T0, T1, T2) as Deserialize<'de>>
             <(T0, T1, T2, T3) as Deserialize<'de>>
             <(T0, T1, T2, T3, T4) as Deserialize<'de>>
           and 820 others
   = note: required for `std::option::Option<Tz>` to implement `Deserialize<'_>`
note: required by a bound in `app_error::_::_serde::__private::de::missing_field`
  --> /home/nange/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.164/src/private/de.rs:22:8
   |
20 | pub fn missing_field<'de, V, E>(field: &'static str) -> Result<V, E>
   |        ------------- required by a bound in this function
21 | where
22 |     V: Deserialize<'de>,
   |        ^^^^^^^^^^^^^^^^ required by this bound in `missing_field`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `common-meta-app` (lib) due to 4 previous errors
warning: build failed, waiting for other jobs to finish...
```

**Comment by ariesdevil** [comment]
Hi, please run `cargo --version` and `rustc --version` under the databend directory, and paste the output here.
And if you enable `hive` feature, please disable it for now, this feature is disabled default.

**Comment by nange** [comment]
image
```
cargo 1.72.0-nightly (b0fa79679 2023-06-03)
rustc 1.72.0-nightly (8b35c0bb0 2023-06-08)
```

I didn't manually enable the `hive` feature.  @ariesdevil

**Comment by ariesdevil** [comment]
Any suggestions for this compile error? @PsiACE @Xuanwo @everpcpc

**Comment by Xuanwo** [comment]
What's your output of `cargo clean && cargo check`?

**Comment by nange** [comment]
> What's your output of `cargo clean && cargo check`?

```
Compiling proc-macro2 v1.0.56
Compiling quote v1.0.26
Compiling unicode-ident v1.0.8
Compiling autocfg v1.1.0
Compiling libc v0.2.140
Compiling syn v1.0.109
Checking cfg-if v1.0.0
Compiling version_check v0.9.4
Compiling serde v1.0.164
...
...
   Checking arrow-csv v40.0.0
    Checking arrow-json v40.0.0
    Checking arrow-ord v40.0.0
    Checking arrow-string v40.0.0
   Compiling enterprise-meta v0.1.0 (/home/nange/rust/github.com/nange/databend/src/meta/ee)
    Checking sqllogictest v0.13.0
    Checking arrow-row v40.0.0
    Checking arrow-arith v40.0.0
error: expected one of `:`, `;`, `=`, `@`, or `|`, found `-`
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:72749:15
      |
72749 |     let mut f_-1: Option<NoSuchObjectException> = None;
      |               ^ expected one of `:`, `;`, `=`, `@`, or `|`

error[E0463]: can't find crate for `ordered_float`
 --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:5:1
  |
5 | extern crate ordered_float;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error[E0463]: can't find crate for `try_from`
 --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:7:1
  |
7 | extern crate try_from;
  | ^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error[E0428]: the name `GetTablesResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:48813:1
      |
16849 | pub struct GetTablesResult {
      | -------------------------- previous definition of the type `GetTablesResult` here
...
48813 | struct GetTablesResult {
      | ^^^^^^^^^^^^^^^^^^^^^^ `GetTablesResult` redefined here
      |
      = note: `GetTablesResult` must be defined only once in the type namespace of this module

error[E0428]: the name `GetTableResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:49583:1
      |
16685 | pub struct GetTableResult {
      | ------------------------- previous definition of the type `GetTableResult` here
...
49583 | struct GetTableResult {
      | ^^^^^^^^^^^^^^^^^^^^^ `GetTableResult` redefined here
      |
      = note: `GetTableResult` must be defined only once in the type namespace of this module

error[E0428]: the name `AddPartitionsResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:51547:1
      |
10061 | pub struct AddPartitionsResult {
      | ------------------------------ previous definition of the type `AddPartitionsResult` here
...
51547 | struct AddPartitionsResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^ `AddPartitionsResult` redefined here
      |
      = note: `AddPartitionsResult` must be defined only once in the type namespace of this module

error[E0428]: the name `GetFileMetadataByExprResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:69280:1
      |
15722 | pub struct GetFileMetadataByExprResult {
      | -------------------------------------- previous definition of the type `GetFileMetadataByExprResult` here
...
69280 | struct GetFileMetadataByExprResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `GetFileMetadataByExprResult` redefined here
      |
      = note: `GetFileMetadataByExprResult` must be defined only once in the type namespace of this module

error[E0428]: the name `GetFileMetadataResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:69394:1
      |
15906 | pub struct GetFileMetadataResult {
      | -------------------------------- previous definition of the type `GetFileMetadataResult` here
...
69394 | struct GetFileMetadataResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `GetFileMetadataResult` redefined here
      |
      = note: `GetFileMetadataResult` must be defined only once in the type namespace of this module

error[E0428]: the name `PutFileMetadataResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:69508:1
      |
16046 | pub struct PutFileMetadataResult {
      | -------------------------------- previous definition of the type `PutFileMetadataResult` here
...
69508 | struct PutFileMetadataResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `PutFileMetadataResult` redefined here
      |
      = note: `PutFileMetadataResult` must be defined only once in the type namespace of this module

error[E0428]: the name `ClearFileMetadataResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:69622:1
      |
16192 | pub struct ClearFileMetadataResult {
      | ---------------------------------- previous definition of the type `ClearFileMetadataResult` here
...
69622 | struct ClearFileMetadataResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `ClearFileMetadataResult` redefined here
      |
      = note: `ClearFileMetadataResult` must be defined only once in the type namespace of this module

error[E0428]: the name `CacheFileMetadataResult` is defined multiple times
     --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:69736:1
      |
16300 | pub struct CacheFileMetadataResult {
      | ---------------------------------- previous definition of the type `CacheFileMetadataResult` here
...
69736 | struct CacheFileMetadataResult {
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `CacheFileMetadataResult` redefined here
      |
      = note: `CacheFileMetadataResult` must be defined only once in the type namespace of this module

    Checking lz4 v1.24.0
    Checking bzip2 v0.4.4
warning: variant `LOAD_DONE` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:107:3
    |
107 |   LOAD_DONE = 1,
    |   ^^^^^^^^^ help: convert the identifier to upper camel case: `LoadDone`
    |
    = note: `#[warn(non_camel_case_types)]` on by default

warning: variant `NOT_ACQUIRED` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:214:3
    |
214 |   NOT_ACQUIRED = 4,
    |   ^^^^^^^^^^^^ help: convert the identifier to upper camel case: `NotAcquired`

warning: variant `SHARED_READ` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:249:3
    |
249 |   SHARED_READ = 1,
    |   ^^^^^^^^^^^ help: convert the identifier to upper camel case: `SharedRead`

warning: variant `SHARED_WRITE` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:250:3
    |
250 |   SHARED_WRITE = 2,
    |   ^^^^^^^^^^^^ help: convert the identifier to upper camel case: `SharedWrite`

warning: variant `NO_TXN` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:358:3
    |
358 |   NO_TXN = 6,
    |   ^^^^^^ help: convert the identifier to upper camel case: `NoTxn`

warning: variant `SCHEMA_REGISTRY` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:432:3
    |
432 |   SCHEMA_REGISTRY = 2,
    |   ^^^^^^^^^^^^^^^ help: convert the identifier to upper camel case: `SchemaRegistry`

warning: variant `START_REVIEW` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:572:3
    |
572 |   START_REVIEW = 2,
    |   ^^^^^^^^^^^^ help: convert the identifier to upper camel case: `StartReview`

warning: variant `CHANGES_REQUIRED` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:573:3
    |
573 |   CHANGES_REQUIRED = 3,
    |   ^^^^^^^^^^^^^^^^ help: convert the identifier to upper camel case: `ChangesRequired`

warning: variant `ORC_SARG` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:685:3
    |
685 |   ORC_SARG = 1,
    |   ^^^^^^^^ help: convert the identifier to upper camel case: `OrcSarg`

warning: variant `TEST_CAPABILITY` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:717:3
    |
717 |   TEST_CAPABILITY = 1,
    |   ^^^^^^^^^^^^^^^ help: convert the identifier to upper camel case: `TestCapability`

warning: variant `INSERT_ONLY_TABLES` should have an upper camel case name
   --> /home/nange/rust/github.com/nange/databend/target/debug/build/common-hive-meta-store-dd2a6ee278ea616a/out/hms_patched.rs:718:3
    |
718 |   INSERT_ONLY_TABLES = 2,
    |   ^^^^^^^^^^^^^^^^^^ help: convert the identifier to upper camel case: `InsertOnlyTables`

    Checking opensrv-mysql v0.3.0 (https://github.com/datafuselabs/opensrv?rev=5e37788#5e377885)
    Checking mysql_async v0.31.3
   Compiling pyo3-macros v0.18.3
Some errors have detailed explanations: E0428, E0463.
For more information about an error, try `rustc --explain E0428`.
warning: `common-hive-meta-store` (lib) generated 11 warnings
error: could not compile `common-hive-meta-store` (lib) due to 11 previous errors; 11 warnings emitted
warning: build failed, waiting for other jobs to finish...
```

@Xuanwo

**Comment by Xuanwo** [comment]
This is quiet confused. `common-hive-meta-store` should not compiled if `hive` feature not enabled. Can you give this the log of `cargo build -v` (could long, please upload as file) and the output of `thrift --version`?

**Comment by nange** [comment]
> This is quiet confused. `common-hive-meta-store` should not compiled if `hive` feature not enabled. Can you give this the log of `cargo build -v` (could be long, please upload as file) and the output of `thrift --version`?

`cargo build -v` logs: [output.log](https://github.com/datafuselabs/databend/files/12116870/output.log)

`thrift --version`: Thrift version 0.11.0

@Xuanwo

**Comment by ariesdevil** [comment]
Please try `bash ./scripts/setup/run_build_tool.sh make test`

**Comment by Xuanwo** [comment]
0.11.0 could be too old, please try upgrade thrift to v0.18 instead.

**Comment by nange** [comment]
Using `bash ./scripts/setup/run_build_tool.sh make test` everything is ok. 
And `cargo check && cargo test` has no errors after upgraded thrift to v0.18.  

So the problem is the thrift version being too low, and `make setup -d` may not upgrade the dependencies to the appropriate version.

Thank you. @ariesdevil @Xuanwo

**Comment by nange** [comment]
Hi, @ariesdevil  It's ready for review. PTAL.

**Review by ariesdevil — APPROVED** [review]
Nice work

**Comment by ariesdevil** [comment]
@soyeric128 please review the doc part

**Comment by BohuTANG** [comment]
Good job, we can merge first then the doc :)

File: docs/doc/15-sql-functions/131-geo-h3-functions/index.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+---
+title: 'Geography H3 Functions'
```
Comment (BohuTANG) [inline docs/doc/15-sql-functions/131-geo-h3-functions/index.md:2]: We already have a GEO doc:https://databend.rs/doc/sql-functions/geo-functions/

It seems no need add a new one.

Comment (nange) [reply docs/doc/15-sql-functions/131-geo-h3-functions/index.md]: Fixed.

**Pull request merged**